File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        
#         ans = []
        
#         def inord(n):
#             if not n:
#                 return
            
#             inord(n.left)
#             ans.append(n.val)
#             inord(n.right)
            
#         inord(root)
#         return ans

        st = deque()
        res = []
        curr = root
            
        while st or curr:
            
            if curr:
                st.append(curr)
                curr = curr.left
            else:
                curr = st.pop()
                res.append(curr.val)
                
                curr = curr.right
                
                
        return res
                
            

File: 0094-binary-tree-inorder-traversal/0094-binary-tree-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        // morris traversal
        
        TreeNode *curr = root;
        vector<int> res;
        
        while(curr){
            // printing leftnode
            if(curr->left==NULL){
                res.push_back(curr->val);
                curr=curr->right;
            }
            else{                
                TreeNode *tmp = curr->left;
                while(tmp->right != NULL && tmp->right != curr)
                    tmp = tmp->right;
                // making new connections
                if(tmp->right==NULL){
                    tmp->right=curr;
                    curr=curr->left;
                }
                //removing those connections
                else{
                    tmp->right=NULL;
                    res.push_back(curr->val);
                    curr=curr->right;
                }
            }
        }
        
        return res;
    }
};